### PR TITLE
Mark C API as dllexports on Windows

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
On Windows, symbols have to be explicitly exported to make them visible/accessible in a shared library. Binaryen.dll currently doesn't export any symbols as a result.

Marking all exported methods as `BINARYEN_API` that is defined as `__declspec(dllexport)` on Windows, unless building a static library.

Sorry for the lots of diffs.

### CODE REVIEW

**Comment by dcodeIO** [comment]
Wondering if, in an Emscripten build, this could be reused to keep the symbols alive, so we don't have to update that [longish list](https://github.com/WebAssembly/binaryen/blob/master/build-js.sh#L182) in build-js.sh anymore?

**Comment by MichalStrehovsky** [comment]
> Wondering if, in an Emscripten build, this could be reused to keep the symbols alive, so we don't have to update that [longish list](https://github.com/WebAssembly/binaryen/blob/master/build-js.sh#L182) in build-js.sh anymore?

Would that be achievable by `#define`-ing the new `BINARYEN_API` macro as `EMSCRIPTEN_KEEPALIVE` when targeting Emscripten? Happy to add that.

File: CMakeLists.txt [hunk 1]
```diff
@@ -192,6 +192,7 @@ SET(binaryen_SOURCES
 )
 IF(BUILD_STATIC_LIB)
   ADD_LIBRARY(binaryen STATIC ${binaryen_SOURCES})
+  ADD_DEFINITIONS(-D BUILD_STATIC_LIBRARY)
```
Comment (sbc100) [inline CMakeLists.txt]: nit: we don't normally include a space after -D

File: src/binaryen-c.cpp [hunk 2]
```diff
@@ -262,16 +262,16 @@ extern "C" {
...
-BinaryenType BinaryenTypeInt32(void) { return i32; }
-BinaryenType BinaryenTypeInt64(void) { return i64; }
-BinaryenType BinaryenTypeFloat32(void) { return f32; }
-BinaryenType BinaryenTypeFloat64(void) { return f64; }
-BinaryenType BinaryenTypeVec128(void) { return v128; }
-BinaryenType BinaryenTypeAnyref(void) { return anyref; }
-BinaryenType BinaryenTypeExnref(void) { return exnref; }
-BinaryenType BinaryenTypeUnreachable(void) { return unreachable; }
-BinaryenType BinaryenTypeAuto(void) { return uint32_t(-1); }
+BINARYEN_API BinaryenType BinaryenTypeNone(void) { return none; }
```
Comment (sbc100) [inline src/binaryen-c.cpp]: Is it not enough to mark these as exported in the header file alone?

**Comment by kripken** [comment]
> Is Binaryen.dll only supposed to export the C API? What about uses of the DLL who want to use the C++ API? 

Hmm, good question! I was really just thinking of the C API myself, and am not aware of any users of the C++ API with libbinaryen. But maybe that could be useful?

**Comment by sbc100** [comment]
OK well lets go with C API only for now.

We can do a follow up to limit the exported symbols on linux too.

**Comment by kripken** [comment]
Sounds good.

I don't know that much about exporting etc. on windows, so I can't review this myself. Does it look good to land to you @sbc100?

**Comment by sbc100** [comment]
Aside from my two comments this seems good yes.

**Comment by MichalStrehovsky** [comment]
I think I addressed all the feedback. Thanks!

**Pull request merged**